File: app/src/main/java/com/flatstack/android/utils/storage/IStorage.java
package com.flatstack.android.utils.storage;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.reflect.Type;
import java.util.List;

/**
 * Created by ereminilya on 8/1/17.
 */

public interface IStorage {

    @Nullable <T> T get(@NonNull String key, @NonNull Type type);

    <T> void put(@NonNull String key, @NonNull T item);

    void putString(@NonNull final String key, @NonNull String str);

    @Nullable String getString(@NonNull String key);

    void putLong(@NonNull String key, long number);

    long getLong(@NonNull String key, long defaultValue);

    void putInt(@NonNull String key, int number);

    int getInt(@NonNull String key, int defaultValue);

    void putBoolean(@NonNull String key, boolean value);

    boolean getBoolean(@NonNull String key, boolean defaultValue);

    void remove(@NonNull String key);

    <T> void putCollection(@NonNull String key, @NonNull List<T> items);

    @Nullable <T> List<T> getCollection(@NonNull String key, @NonNull Type type);
}

File: app/src/main/java/com/flatstack/android/util/storage/IStorage.kt
package com.flatstack.android.util.storage

import java.lang.reflect.Type

@Suppress("TooManyFunctions")
interface IStorage {

    operator fun <T> get(key: String, type: Type): T?

    fun put(key: String, items: Any)

    fun putString(key: String, str: String)

    fun getString(key: String): String?

    fun putLong(key: String, number: Long)

    fun getLong(key: String, defaultValue: Long): Long

    fun putInt(key: String, number: Int)

    fun getInt(key: String, defaultValue: Int): Int

    fun putBoolean(key: String, value: Boolean)

    fun getBoolean(key: String, defaultValue: Boolean): Boolean

    fun remove(key: String)

    fun <T> putCollection(key: String, items: List<T>)

    fun <T> getCollection(key: String, type: Type): List<T>?
}
